File: scripts/current_user.py
import ipaddress
import json
import os


def write_json(payload):
    print(json.dumps(payload, ensure_ascii=False))


def normalize_remote_ip(value):
    ip_addr = str(value or "").split(",", 1)[0].strip()
    if ip_addr.startswith("::ffff:"):
        ip_addr = ip_addr[7:]
    return ip_addr


def main():
    ip_addr = normalize_remote_ip(os.environ.get("REMOTE_ADDR"))
    try:
        normalized_ip = str(ipaddress.ip_address(ip_addr))
    except ValueError:
        write_json({
            "ok": False,
            "code": "IP_NOT_FOUND",
            "error": "Unable to determine the client IP.",
        })
        return

    write_json({"ok": True, "knoxId": normalized_ip})


if __name__ == "__main__":
    main()

File: server/currentUser.mjs
import { isIP } from "node:net"

import { createSafeApiError } from "./safeApiError.mjs"

function sendJson(res, statusCode, payload) {
  res.writeHead(statusCode, {
    "Content-Type": "application/json; charset=utf-8",
    "Cache-Control": "no-store",
  })
  res.end(JSON.stringify(payload))
}

export function normalizeRemoteIp(value) {
  const ip = String(value ?? "").split(",")[0].trim()
  return ip.startsWith("::ffff:") ? ip.slice(7) : ip
}

export function getRemoteIp(req) {
  const remoteIp = normalizeRemoteIp(
    req.headers["x-forwarded-for"]
    ?? req.headers["x-real-ip"]
    ?? req.socket?.remoteAddress
    ?? "",
  )
  return isIP(remoteIp) ? remoteIp : ""
}

export async function resolveCurrentUser(remoteIp) {
  const normalizedRemoteIp = normalizeRemoteIp(remoteIp)
  if (!isIP(normalizedRemoteIp)) {
    const error = new Error("접속자 IP를 확인하지 못했습니다.")
    error.code = "IP_NOT_FOUND"
    throw error
  }
  return { ok: true, knoxId: normalizedRemoteIp }
}

export async function handleCurrentUserRequest(req, res) {
  if (req.method !== "GET") {
    sendJson(res, 405, { ok: false, error: "Method not allowed" })
    return
  }

  const remoteIp = getRemoteIp(req)
  if (!remoteIp) {
    sendJson(res, 400, { ok: false, error: "접속자 IP를 확인하지 못했습니다." })
    return
  }

  try {
    const payload = await resolveCurrentUser(remoteIp)
    sendJson(res, 200, payload)
  } catch {
    sendJson(res, 400, createSafeApiError({
      code: "CURRENT_USER_IP_INVALID",
      message: "접속자 IP를 확인하지 못했습니다.",
      scope: "current-user",
    }))
  }
}
